clients/go/compat.go

```diff
@@ -228,40 +249,60 @@ func (b WrappedBackend) SuggestGasTipCap(ctx context.Context) (*big.Int, error)
 
 // EstimateGas implements ContractTransactor.
 func (b WrappedBackend) EstimateGas(ctx context.Context, call ethereum.CallMsg) (gas uint64, err error) {
-	// Assume latest round in oasis-web3-gateway RPC call
-	return b.backend.EstimateGas(ctx, call)
+	var packedCall *ethereum.CallMsg
+	if call.From == [common.AddressLength]byte{} {
+		var err error
+		if packedCall, err = PackCall(call, b.cipher); err != nil {
+			return 0, err
+		}
+	} else {
+		leash, err := b.makeLeash(ctx, call.From, nil)
+		if err != nil {
+			return 0, err
+		}
+		if packedCall, err = PackSignedCall(call, b.cipher, b.sign, b.chainID, leash); err != nil {
+			return 0, fmt.Errorf("failed to pack signed call: %w", err)
+		}
+	}
+
+	return b.backend.EstimateGas(ctx, *packedCall)
+}
+
+// makeLeash creates a new leash for the given from address and blockNumber.
+// If blockNumber is nil, the latest block is taken.
+func (b WrappedBackend) makeLeash(ctx context.Context, from common.Address, blockNumber *big.Int) (*evm.Leash, error) {
+	leashBlockNumber := big.NewInt(0)
+	header, err := b.backend.HeaderByNumber(ctx, blockNumber) // NB: blockNumber==nil will fetch the latest block.
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch leash block header: %w", err)
+	}
+	// We will build a leash on the pre-last block.
+	blockHash := header.ParentHash
+	leashBlockNumber.Sub(header.Number, big.NewInt(1))
+	nonce, err := b.backend.PendingNonceAt(ctx, from)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch account nonce: %w", err)
+	}
+	return &evm.Leash{
+		Nonce:       nonce,
+		BlockNumber: leashBlockNumber.Uint64(),
+		BlockHash:   blockHash[:],
+		BlockRange:  DefaultBlockRange,
+	}, nil
 }
 
 func txNeedsPacking(tx *types.Transaction) bool {
 	_, isPacked := packedTxs.Load(tx.Hash())
```

Comment:
Why do we need `packedTxs`? Can't we infer that from the transaction itself? This seems hacky and also a potential memory leak with `packedTxs` not getting cleaned up.